libraries/standard/http/src/http_client.c

```diff
@@ -155,21 +155,14 @@ static HTTPStatus_t _addHeader( HTTPRequestHeaders_t * pRequestHeaders,
                                  ( int32_t ) fieldLen, pField,
                                  ( int32_t ) valueLen, pValue );
 
-        if( ( bytesWritten + HTTP_HEADER_LINE_SEPARATOR_LEN ) != toAddLen )
-        {
-            IotLogErrorWithArgs( "Internal error in snprintf() in _addHeader(). "
-                                 "BytesWritten: %d.", bytesWritten );
-        }
-        else
-        {
-            pBufferCur += bytesWritten;
+        assert( bytesWritten == (int) ( toAddLen - HTTP_HEADER_LINE_SEPARATOR_LEN ) );
```

Comment:
We are not asserting here before the input into snprintf is different than the integers we were writing in AddRangeHeader().
The inputs into snprintf here are from user inputs, so we need to bubble up the encoding error.

Reply:
The `snprintf` is only called after the buffer space is validated.

Reply:
It is not about the space, it's the encoding error from invalid characters that cannot be encoded

Reply:
If the worry is about non-printable characters, I have tested with the following on gcc:
```
#include <stdio.h>
#include <assert.h>
#include <stdint.h>

int main()
{
    char buffer[ 100 ];
    uint8_t input[] = { 0xFF, 0xFF };

    assert( snprintf( buffer, 100, "%.*s", 2, ( char * ) input ) == 2 );
}
```

Reply:
This code is to be portable to non-gcc compilers

Reply:
If the goal is to propagate the error back to the user, then an `INTERNAL_FAILURE` error code cannot be helpful. (Also, it seems like the original code didn't return any error code for the failure: https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/development/libraries/standard/http/src/http_client.c#L158-L162)

Ideally, we should provide `INVALID_PARAMETER` back to the caller, but there isn't a reliable way to do without actually doing O(n) traversal on both the header and value string inputs

Reply:
That is a miss in the original code, but there was supposed to be an error returned.
(continuing conversation offline)

Reply:
We discussed offline, and decided to use `uint8_t *` instead of `char *` for user-input as the HTTP standard supports [ISO 8859-1]() character encoding which has a larger range set that `char` type.
Also, to avoid any platform-specific `snprintf` issues with printing characters outside the `char` range, we will use `memcpy` to just copy the bytes into the buffer.